### PR TITLE
Feature/crud categories

### PR DESCRIPTION
**Pull request by K4sku**
Sprint 6 / task 3: crud-for-categories

PR changes: 
- add routes for Category
- add Category model: validates titles presence and uniqueness
- add Category controller
- add views for Category
- add CategoryReward model (join table)
- changes to Reward model: add category association, validate presence of category
- changes to Reward controller: allow category in params and pass possible_categories (list of unselected categories) to view
- changes to Reward views: add form field to select categories, add display of categories for Reward, fix back link from show action
- add migrations for associations
- update seeds for rewards with categories
- add factories, Category model spec, specs for creating and deleting categories, specs for adding and removing categories from rewards
- add Category link to admins navbar panel
- add frontend dependecy "Bulma-TagsInput" that renders category selector
- import and call Bulma-TagsInput JS

category_rewards

PS. sorry for size of PR, some refactors for specs were necessary to accommodate rewards requiring presence of category

### CODE REVIEW

File: app/controllers/admins/rewards_controller.rb [hunk 1]
```diff
@@ -10,17 +10,20 @@ def show
 
     def new
       @reward = Reward.new
+      @possible_categories = Category.all
```
Comment (K4sku) [inline app/controllers/admins/rewards_controller.rb:13]: @possible_categories is required for bulma form element to properly display possible options

File: app/views/admins/rewards/_form.html.erb [hunk 2]
```diff
@@ -4,6 +4,28 @@
   <%= form.text_field :description %>
   <%= form.number_field :price, min: 1, step: 1 %>
 
+  <input name="reward[category_ids][]" type="hidden" value>
```
Comment (K4sku) [inline app/views/admins/rewards/_form.html.erb:7]: Bulma extention TagsImput is used here. 

Hidden input is there to send empty array when all categories are deleted (multiselect form helper in rails does this).

File: app/views/admins/rewards/_form.html.erb [hunk 3]
```diff
@@ -4,6 +4,28 @@
...
+  <div class="field">
+    <%= form.label :category_ids, "Categories", class: 'label' %>
+    <div class="control">
+      <select id="reward_category_ids" name ="reward[category_ids][]" multiple data-type="tags" data-placeholder="Choose Categories" test_id='category_input'>
+        <%= @reward.categories.each do |category| %>
+          <option value="<%= category.id %>" selected><%= category.title %></option>
+        <% end %>
+        <%= @possible_categories.each do |possible_category| %>
```
Comment (K4sku) [inline app/views/admins/rewards/_form.html.erb:15]: adds categories to selection list

File: app/controllers/admins/categories_controller.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    def update
+      if category.update(category_params)
+        redirect_to admins_category_path(category), notice: "Category #{category.title} was updated"
+      else
+        render :edit
+      end
+    end
+
+    def destroy
+      if category.rewards.any?
```
Comment (K4sku) [inline app/controllers/admins/categories_controller.rb]: prevent destroying category with rewards

File: db/migrate/20220814093999_update_reward_categories_where_null.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class UpdateRewardCategoriesWhereNull < ActiveRecord::Migration[6.1]
```
Comment (SzymonNadbrzezny) [inline db/migrate/20220814093999_update_reward_categories_where_null.rb]: If you want to do data migration consider placing  it in separate file like `db/data_migrations/` and creating rake tasks for them.

Visuality's article about it: https://www.visuality.pl/posts/safe-data-migrations-in-rails

File: db/migrate/20220814094218_add_not_null_contraint_on_category_title.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddNotNullContraintOnCategoryTitle < ActiveRecord::Migration[6.1]
```
Comment (SzymonNadbrzezny) [inline db/migrate/20220814094218_add_not_null_contraint_on_category_title.rb]: You could cleanup migrations and edit old ones to create fields that aren't nullable. I belive it would be more readable than creating multiple migrations to edit single field in the same PR.

File: spec/factories/factory_sequences.rb [hunk 7]
```diff
@@ -8,4 +8,7 @@
   sequence :company_value_title do |n|
     "CValue#{n}"
   end
+  sequence :category_title do |n|
+    "CValue#{n}"
```
Comment (SzymonNadbrzezny) [inline spec/factories/factory_sequences.rb]: I think it would be better to use different value for categories and company values for readability. Perhaps using model names in the titles?

File: spec/factories/employee_factory.rb [hunk 8]
```diff
@@ -3,5 +3,9 @@
     email { generate :employee_email }
     password { 'password' }
     number_of_available_kudos { 10 }
+
+    trait :with_point do
```
Comment (SzymonNadbrzezny) [inline spec/factories/employee_factory.rb]: This is outside of AC. It'd be best to move it to separate PR.

File: app/controllers/admins/categories_controller.rb (hunk 4 above)
Comment (SzymonNadbrzezny) [reply app/controllers/admins/categories_controller.rb]: Maybe you could use Policy Object for that?

**Review by SzymonNadbrzezny — COMMENTED** [review]
Looks good 👍  , left some suggestions.

File: db/migrate/20220813065131_create_category_rewards.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+class CreateCategoryRewards < ActiveRecord::Migration[6.1]
+  def change
+    create_table :category_rewards do |t|
+      t.references :category, index: true
+      t.references :reward, index: true
```
Comment (maikhel) [inline db/migrate/20220813065131_create_category_rewards.rb:5]: You could add `t.timestamps` to keep it consistent with other tables (in Rails usually each table have created_at, updated_at columns). Also, maybe such columns would be useful for debugging  later on.

File: db/migrate/20220814094218_add_not_null_contraint_on_category_title.rb (hunk 6 above)
Comment (maikhel) [reply db/migrate/20220814094218_add_not_null_contraint_on_category_title.rb]: Agree, looks like too many migrations for such a simple feature.

File: db/migrate/20220814093999_update_reward_categories_where_null.rb (hunk 5 above)
Comment (maikhel) [reply db/migrate/20220814093999_update_reward_categories_where_null.rb]: Agree. But that's so cool you remembered about updating existing records!

File: app/views/admins/categories/_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+<%= bulma_form_with(model: [:admins, category]) do |form| %>
+  <%= form.text_field :title, autofocus: true %>
+
+  <div class="actions">
+    <%= form.submit "Save Categotry", class: "button is-primary" %>
```
Comment (maikhel) [inline app/views/admins/categories/_form.html.erb]: typo, "Category"

File: spec/system/admins/caregories/destroy_spec.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
+require 'rails_helper'
+
+describe 'admins destroy Category', type: :system, js: true do
```
Comment (maikhel) [inline spec/system/admins/caregories/destroy_spec.rb]: typo in filename, `categories/`

File: spec/system/admins/caregories/destroy_spec.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      expect(page).not_to have_selector(:css, "div[test_id='category_#{category.id}']")
+    end
+  end
+
+  context 'when admin is destroying category with rewards' do
+    it 'destroys category record' do
+      category_with_rewards = create(:category_with_rewards)
+      visit current_path
+      expect(page).to have_selector(:css, "div[test_id='category_#{category_with_rewards.id}']")
+      within(page.find(:css, "div[test_id='category_#{category_with_rewards.id}']")) do
```
Comment (maikhel) [inline spec/system/admins/caregories/destroy_spec.rb]: If you use css selector in many places, you can store it in variable, e. g.
```
category_element = "div[test_id='category_#{category_with_rewards.id}']"
```

File: spec/system/admins/caregories/destroy_spec.rb [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+      accept_alert do
+        click_on 'Delete'
+      end
+      expect(page).to have_current_path admins_categories_path
+      expect(page).not_to have_selector(:css, "div[test_id='category_#{category.id}']")
+    end
+  end
+
+  context 'when admin is destroying category with rewards' do
+    it 'destroys category record' do
```
Comment (maikhel) [inline spec/system/admins/caregories/destroy_spec.rb]: Actually, `it 'prevents from destroying category'`

File: spec/factories/factory_sequences.rb (hunk 7 above)
Comment (maikhel) [reply spec/factories/factory_sequences.rb]: True, or you can use `faker` gem to generate random strings.

File: db/seeds.rb [hunk 14]
```diff
@@ -26,13 +26,19 @@
   end
 end
 
+Category.create!([
```
Comment (maikhel) [inline db/seeds.rb:29]: Cool you updated seeds! 👍

File: app/views/admins/categories/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    <div class="level-left">
+      <p class='is-size-3'><strong>Categories</strong></p>
+    </div>
+    <div class="level-right">
+      <%= link_to 'New Category', new_admins_category_path, class: 'button is-primary level-item' %>
+    </div>
+  </nav>
+  <div class='box'>
+    <div class="list has-hoverable-list-items has-visible-pointer-controls">
+      <% @categories.each do |category| %>
```
Comment (maikhel) [inline app/views/admins/categories/index.html.erb:12]: maybe `includes(:rewards)` would be useful here? There is `category.rewards.count` which can cause n+1 query problems

**Review by maikhel — APPROVED** [review]
Looks good!

File: app/views/admins/categories/index.html.erb (hunk 15 above)
Comment (K4sku) [reply app/views/admins/categories/index.html.erb:12]: Good catch, thank you! 
includes didn't do the job, however I managed to query rewards_count into a hash. @maikhel Do you think this is a good solution here?

File: spec/factories/factory_sequences.rb (hunk 7 above)
Comment (K4sku) [reply spec/factories/factory_sequences.rb]: Thanks for pointing this out. Fixed.

File: db/migrate/20220814094218_add_not_null_contraint_on_category_title.rb (hunk 6 above)
Comment (K4sku) [reply db/migrate/20220814094218_add_not_null_contraint_on_category_title.rb]: I have merged migrations to 3 files: create categories, creating category_rewards and  adding 'not null' to category_rewards (it has to be separate to create space for running data migration) 
Data migration is moved to separate folder with rake task.

File: spec/system/admins/caregories/destroy_spec.rb (hunk 12 above)
Comment (K4sku) [reply spec/system/admins/caregories/destroy_spec.rb]: That's cool, did not know that. Thank you!

**Review by K4sku — COMMENTED** [review]
Thank you for your suggestions 👍
